engine/flink/model-util/src/test/scala/pl/touk/nussknacker/engine/flink/util/transformer/aggregate/OuterJoinTransformerSpec.scala

```diff
@@ -0,0 +1,190 @@
+      .filter(_.variableTyped(KeyVariableName).contains(key))
+      .map(_.variableTyped[java.lang.Integer](OutVariableName).get)
+
+    outValues shouldEqual List(null, 123)
+  }
+
+  private def runProcess(testProcess: EspProcess, input1: List[OneRecord], input2: List[OneRecord]): TestResults[Any] = {
+    val collectingListener = ResultsCollectingListenerHolder.registerRun(identity)
+    val model = modelData(input1, input2, collectingListener)
+    val stoppableEnv = StoppableExecutionEnvironment(FlinkTestConfiguration.configuration())
+    try {
+      val registrar = FlinkStreamingProcessRegistrar(new FlinkProcessCompiler(model), model.processConfig, ExecutionConfigPreparer.unOptimizedChain(model, None))
+      registrar.register(new StreamExecutionEnvironment(stoppableEnv), testProcess, ProcessVersion.empty, Some(collectingListener.runId))
+      val id = stoppableEnv.execute(testProcess.id)
+      stoppableEnv.waitForJobState(id.getJobID, testProcess.id, ExecutionState.FINISHED)()
+      collectingListener.results[Any]
+    } finally {
+      stoppableEnv.stop()
+    }
+  }
+
+  private def modelData(input1: List[OneRecord], input2: List[OneRecord], collectingListener: ResultsCollectingListener) =
+    LocalModelData(ConfigFactory.empty(), new OuterJoinTransformerSpec.Creator(input1, input2, collectingListener))
+
+}
+
+object OuterJoinTransformerSpec {
+
+  class Creator(mainRecords: List[OneRecord], joinedRecords: List[OneRecord], collectingListener: ResultsCollectingListener) extends EmptyProcessConfigCreator {
+
+    override def customStreamTransformers(processObjectDependencies: ProcessObjectDependencies): Map[String, WithCategories[CustomStreamTransformer]] =
+      Map(
+        "outer-join" -> WithCategories(new OuterJoinTransformer(None) {
+          // We can't be sure that each main records will be consumed after matching joined records - main stream is consumed instantly.
+          // But thanks to fact, that sliding aggregates handles out of order events, we just need to ensure that all joined records are consumed before main records.
+          override protected def prepareAggregatorFunction(aggregator: Aggregator, stateTimeout: FiniteDuration)(implicit nodeId: ProcessCompilationError.NodeId):
+          CoProcessFunction[ValueWithContext[String], ValueWithContext[StringKeyedValue[AnyRef]], ValueWithContext[AnyRef]] = {
+            val joinedSize = joinedRecords.size
+            new CoProcessFunctionInterceptor(super.prepareAggregatorFunction(aggregator, stateTimeout)) with LazyLogging {
+
+              @transient
+              private lazy val addedToStateCount = new AtomicInteger(0)
+
+              @transient
+              private lazy val stateLock = new Object
+
+              override protected def beforeProcessElement1(value: ValueWithContext[String]): Unit = {
+                logger.info(s"OuterJoin: processing element1 with key: ${value.value} - waiting on all elements added to state...")
```

Comment:
logger.debug? here and in following occurrences?

Reply:
The problem is that I'm not 100% sure, that it will work correctly. I've tried a couple of methods to do such synchronization and I'm not sure if the final one is working. Can we leave it as it is for some time? I have quite similar concept in other project, but there I have kafka source and I think that with collection source there can be some sequence that will cause deadlock ... :-/

Reply:
well... ok... ;)